### PR TITLE
Add regulator driver infrastructure

### PR DESCRIPTION
**Pull request by pabigot**
To support runtime power management we need to be able to control voltage/current regulators.  Although there are boards in-tree that use these, there is no devicetree support for describing them.  This PR provides support based on Linux bindings.

The API is based on [the Linux API](https://www.kernel.org/doc/html/latest/driver-api/regulator.html) but adapted for Zephyr.  The enable operation is asynchronous as it may take time to turn a regulator on.  The disable operation is synchronous as the caller doesn't need to know when the regulator transitions to off.  (Power management infrastructure will, but that's not part of this API.)

On the theory that user space never needs to control system power the API requires supervisor privileges.  This may be changed based on feedback.  The API is also isr-ok and pre-kernel-ok.

A significant gap is the lack of any ability to configure the kernel init level and priority on a per-device level.  Without that this infrastructure is fragile.

The thingy52_nrf52832 board has been updated to provide its two external power rails, but the pseudo-devices that currently control them cannot be replaced until device-specific init priority is supported.

### Commit Contents

* Introduction of standard devicetree properties for power supply control
> d306458c15 devicetree: add properties for power supply control

* Primary regulator commits
> 0afd91e3bd drivers: add infrastructure for regulator devices
> d34246f19e drivers: regulator: add GPIO-controlled regulator driver
> c1bc4515f6 doc: introduce the regulator driver API
> 1724be2c7b tests: drivers: regulator: add test for fixed regulator

* Add Thingy:52 on-board regulator devices
> 85c792590c boards: thingy52_nrf52832: add nodes for on-board voltage regulators

### Blockers for merge:
* [x] merge #27443
* [x] update to actually do asynchronous work
* [x] tests
* [x] merge #27719
* [ ] determine whether this is usable without device-specific init priorities
* [x] documentation on tests and how to select regulator implementation: onoff, sync, or dedicated `power-supply`

### CODE REVIEW

**Comment by pabigot** [comment]
My current thought is to add a `power-supply` optional property to the SPI and I2C device base binding definitions, in lieu of having to add those to every driver.  To be discussed.

**Comment by pabigot** [comment]
The latest push reveals another gap in dependency information: all regulators are included in the image even if only one (or no) device that depends on them is enabled.

Yes, that could be addressed by marking some or all of them `status = "disabled";` but that's pretty painful from an application perspective.

File: drivers/regulator/regulator_gpio.c
Comment (wentongwu) [inline drivers/regulator/regulator_gpio.c:45]: I'm not sure this can happen in ISR.

File: include/drivers/regulator.h
Comment (wentongwu) [inline include/drivers/regulator.h:36]: so caller has to spin for the result like the init function `regulator_gpio_init` does?

File: drivers/regulator/regulator_gpio.c
Comment (wentongwu) [reply drivers/regulator/regulator_gpio.c:45]: from device runtime pm framework perspective, we can't know in advance whether the power state transition of individual device is ISR friendly or not, so IMHO maybe we can't say the onoff_request can be called in ISR.

File: include/drivers/regulator.h
Comment (pabigot) [reply include/drivers/regulator.h:36]: No; notification can be spin-wait, callback, or semaphore; everything supported by [sys_notify](https://docs.zephyrproject.org/latest/reference/misc/notify.html#async-notification).  Caller's choice.

File: drivers/regulator/regulator_gpio.c
Comment (pabigot) [reply drivers/regulator/regulator_gpio.c:45]: `k_busy_wait()` is callable from pre-kernel and interrupt contexts, it's just not generally a good idea because it slows things down.

By design and implementation `onoff_request()` can always be called from an ISR.  It's the responsibility of the service's `start` transition function to guarantee that they can be invoked from that context and, if they need to do something that isn't allowed, return OK after putting something on a work queue or handing the operation off to another thread to complete the operation.  Similarly for `onoff_release()` even though that's a synchronous operation.

Here I still need to do that: the `start` and `stop` operations need to check whether `gpio_pin_set()` return `-EWOULDBLOCK`, as they would on an I2C-based IO extender.  That wouldn't affect the regulator API, though.

For the init the code is incomplete: it could check `k_is_pre_kernel()` to determine whether threads are available, and whether the operation completed immediately.  Depending on context the init would either fail, or could proceed and use a semaphore.  (Though It's a little weird to sleep during driver initialization.  I'm not sure the kernel's designed to have multiple threads running (and accessing drivers) at that point, but as long as only the idle thread is available it should work.)

We don't have a plan/API design for turning devices on and off as required by power management or #19448 without power management, so `k_busy_wait()` is fine in the init function ignoring power optimization.

Comment (wentongwu) [reply drivers/regulator/regulator_gpio.c:45]: > We don't have a plan/API design for turning devices on and off as required by power management

@pabigot I think we need a draft PR first for device runtime PM based on onoff service to do the decision

**Comment by galak** [comment]
Should we be using `regulator-fixed` binding/compat instead of `regulator-gpio`?

**Comment by carlescufi** [comment]
API meeting 4th August 2020:

- Initial draft takes the smallest subset possible from the Linux functionality (enable/disable for GPIO-controlled regulators only)
- Can be later extended to support things like PMICs and other types of regulators

**Comment by galak** [comment]
Looks like `regulator-name` is a required property.  So thinking we should as well, and try and have `label` and `regulator-name` match for Zephyr's purposes at this point.

File: drivers/regulator/regulator_gpio.c
Comment (dcpleung) [inline drivers/regulator/regulator_gpio.c:17]: Should these be on different bit positions? Or that "always on" needs to have 2 bits (one for itself, and the other overlapping "boot_on")?

File: drivers/regulator/regulator_gpio.c
Comment (dcpleung) [inline drivers/regulator/regulator_gpio.c:91]: With "always on" bit set, should the driver take care of returning error? Or can it be handled in `disable()` instead?

Comment (pabigot) [reply drivers/regulator/regulator_gpio.c:91]: For now I'm going to do it in the driver; if we get more drivers then it could be lifted to the wrapper.

**Comment by pabigot** [comment]
> Looks like `regulator-name` is a required property. So thinking we should as well, and try and have `label` and `regulator-name` match for Zephyr's purposes at this point.

I'm going to add `regulator-name` (though it'll be ignored in the driver), but don't agree about documenting that the Zephyr-specific label must match the regulator name.  That would limit proper use of the `label` property in the future, when we are free to make this a human-readable description of the node rather than the globally-unique string used in `device_get_binding()`.

**Comment by chrta** [comment]
Thank you, this is a good addition also for all SiLabs boards, since nearly every peripheral on the board must be enabled with a dedicated io pin.

> My current thought is to add a `power-supply` optional property to the SPI and I2C device base binding definitions, in lieu of having to add those to every driver. To be discussed.

I would prefer if there would be a more general approach, but i don't know how to implement it properly, maybe move `power-supply` optional property to some base binding.

Use cases:

For (nearly all) SiLabs dev boards, there is a virtual serial port (via on-board Segger J-Link, named Board controller in source code) that must be enabled with a gpio pin, see https://github.com/zephyrproject-rtos/zephyr/blob/master/boards/arm/efm32gg_stk3701a/board.c#L20

Also the ethernet phy must be enabled via gpio pin a few lines further down, https://github.com/zephyrproject-rtos/zephyr/blob/master/boards/arm/efm32gg_stk3701a/board.c#L38

**Comment by pabigot** [comment]
Yes, I have the sltb004a.  It motivates a slightly different approach: there are many cases, these included, where we want a more efficient solution that doesn't involve a full on-off service that supports delayed startup and shutdown.  Just controlling a GPIO based on whether a counter is zero or positive handles it.  I'm thinking of having another compatible `regulator-fixed-direct`(?) to handle that case in the same driver.  (Alternatively that optimized variant could be selected for an instance based on other properties, like all delays being zero and the referenced pin being directly controlled by a SOC GPIO peripheral.)  I'm going to take a look at that next.

If there's only one thing that will ever control the signal, and it already knows the state without a counter, I don't think a full regulator API is appropriate.  For that application using `power-gpios` instead of `power-supply` might be better.  Which is probably the case for the sltb004a peripherals.

It should be possible to eliminate `boards/arm/efm32gg_stk3701a/board.h` by putting that information into a `board` node in the devicetree, e.g. `eth-power-enable-gpios`, `eth-ref-clk-gpios` and `eth-reset-gpios` properties.  Not sure how to handle the clock, but it should be possible.  Whether it's worth doing is a maintainer decision.

**Comment by chrta** [comment]
> If there's only one thing that will ever control the signal, and it already knows the state without a counter, I don't think a full regulator API is appropriate. For that application using power-gpios instead of power-supply might be better. Which is probably the case for the sltb004a peripherals.

I agree, that would be better.

File: drivers/regulator/Kconfig.gpio
Comment (gmarull) [inline drivers/regulator/Kconfig.gpio:1]: File should likely be renamed `.fixed` ?

File: drivers/regulator/regulator_fixed.c
Comment (gmarull) [inline drivers/regulator/regulator_fixed.c:127]: `-ENODEV`?

File: drivers/regulator/regulator_fixed.c
Comment (gmarull) [inline drivers/regulator/regulator_fixed.c:111]: could use `=  ONOFF_TRANSITIONS_INITIALIZER(start, stop, NULL)`

File: drivers/regulator/regulator_fixed.c
Comment (gmarull) [inline drivers/regulator/regulator_fixed.c:100]: What is the reason for this check?

File: drivers/regulator/regulator_fixed.c
Comment (gmarull) [inline drivers/regulator/regulator_fixed.c:49]: What is the reason for the last check, i.e. `k_us_to_ticks_floor32(cfg->delay_us) == 0)`?

**Review by gmarull — COMMENTED** [review]
Overall LGTM as an starting point. Left some questions

File: drivers/regulator/regulator_fixed.c
Comment (pabigot) [reply drivers/regulator/regulator_fixed.c:100]: It was intended to verify that the device passed in is a regulator that can be controlled by this driver, something Zephyr doesn't generally do.  In fact it isn't useful here because the driver's API pointer was already used to get here so it must be correct.  Removed.

File: drivers/regulator/regulator_fixed.c
Comment (pabigot) [reply drivers/regulator/regulator_fixed.c:49]: If the delay is less than one tick then `k_busy_wait` is required to avoid waiting longer than necessary, because sleep durations are always a multiple of ticks.  If it's at least one tick then `k_usleep()` would save power.

This area needs rework for other reasons.

**Comment by pabigot** [comment]
Updated by allowing the driver to be told that a particular GPIO doesn't require asynchronous transition support, which eliminates the overhead of the onoff manager for that driver without affecting the caller API.  See `vdd-pwr` in the thingy52 devicetree, and #27443 on which this now depends.

File: drivers/regulator/regulator_fixed.c
Comment (nordic-krch) [inline drivers/regulator/regulator_fixed.c:219]: shouldn't it be `rc == 1` so that only the last disabling user is cleaning the gpio?

Comment (pabigot) [reply drivers/regulator/regulator_fixed.c]: Reworked, thanks.

**Comment by sjg20** [comment]
Peter, any thoughts on the questions I have above? I'll hold off taking another look for now as they were the main things I was unsure about last time.

**Comment by pabigot** [comment]
@sjg20 Sorry, I can't find any comments from you on this PR.  Can you give me a link to them?

**Comment by sjg20** [comment]
@pabigot Sorry I seem to have missed a step so that I could see them but not you. Can you see them now?

File: include/drivers/regulator.h
Comment (pabigot) [reply include/drivers/regulator.h]: Because you may not be enabling and disabling the operator from thread mode, and the transition might take a long time.

File: include/drivers/regulator.h
Comment (pabigot) [reply include/drivers/regulator.h:49]: It means that the regulator is "on".  What that means is not directly, but I would expect it to mean "stable" or "ready to use".

File: include/drivers/regulator.h
Comment (pabigot) [reply include/drivers/regulator.h:76]: No.  This was a design trade-off in the context of developing the on-off service.  If it's necessary to detect when the transition completes that can be done in other ways.

I suspect it will be necessary, so this may have to change.

File: include/drivers/regulator.h
Comment (sjg20) [reply include/drivers/regulator.h]: So where is that documented? Is this implied in some way? Does the 'transition' just mean telling the hardware to turn on the power supply, not actually waiting for it to happen?

I have a feeling you have a lot of documentation somewhere else that I am missing.

Comment (pabigot) [reply include/drivers/regulator.h]: https://docs.zephyrproject.org/latest/reference/resource_management/index.html#resource-management

File: include/drivers/regulator.h
Comment (sjg20) [reply include/drivers/regulator.h:76]: In general you can only know that you have started to turn it off. Of course you can look at datasheets and some of them will have a 'ready' bit which tells you that the power is stable (or fully off).

IMO it is probably enough to know that we have completed 'asking' for it to turn off. A client could add a udelay if needed.

For turning on, I think we need to know that the power is fully stable, since until it is we cannot advance.

File: include/drivers/regulator.h
Comment (pabigot) [reply include/drivers/regulator.h]: Also the regulator documentation in this PR.

File: include/drivers/regulator.h
Comment (sjg20) [reply include/drivers/regulator.h:49]: That sounds good. OK then I think the comment here should be beefed up

File: include/drivers/regulator.h
Comment (sjg20) [reply include/drivers/regulator.h:33]: As above, this can be called from interrupt context

File: include/drivers/regulator.h
Comment (pabigot) [reply include/drivers/regulator.h:52]: Updated; please revisit.

**Comment by sjg20** [comment]
> The latest push reveals another gap in dependency information: all regulators are included in the image even if only one (or no) device that depends on them is enabled.
> 
> Yes, that could be addressed by marking some or all of them `status = "disabled";` but that's pretty painful from an application perspective.

That seems OK to me for now

File: drivers/regulator/regulator_fixed.c
Comment (sjg20) [inline drivers/regulator/regulator_fixed.c:284]: struct comment?

File: drivers/regulator/regulator_fixed.c
Comment (sjg20) [inline drivers/regulator/regulator_fixed.c:335]: What are these three lines for?

File: doc/reference/peripherals/regulators.rst
Comment (sjg20) [inline doc/reference/peripherals/regulators.rst:27]: node

File: tests/drivers/regulator/fixed/README.txt
Comment (sjg20) [inline tests/drivers/regulator/fixed/README.txt:26]: line length?

File: tests/drivers/regulator/fixed/src/main.c
Comment (sjg20) [inline tests/drivers/regulator/fixed/src/main.c:280]: I feel that having blank lines after each comment is a big much. It's a matter of taste I suppose. I'll get a taller monitor :-)

File: boards/arm/thingy52_nrf52832/thingy52_nrf52832.dts
Comment (sjg20) [inline boards/arm/thingy52_nrf52832/thingy52_nrf52832.dts:38]: Why are these commented out?

**Review by sjg20 — APPROVED** [review]

File: boards/arm/thingy52_nrf52832/thingy52_nrf52832.dts
Comment (pabigot) [reply boards/arm/thingy52_nrf52832/thingy52_nrf52832.dts:38]: Because Zephyr's devicetree binding language doesn't allow inclusion of `power.yaml` to extend the [child-binding](https://github.com/zephyrproject-rtos/zephyr/blob/762b806c3d4bf2f1004052558141913f33462cfa/dts/bindings/gpio/gpio-leds.yaml#L8).  That will have to be fixed, but only when this starts to be usable which requires device dependency.

File: drivers/regulator/regulator_fixed.c
Comment (pabigot) [reply drivers/regulator/regulator_fixed.c:284]: I suppose we can start requiring that for internal implementation data structures that aren't processed to provide documentation.  Do we want to do that?

File: drivers/regulator/regulator_fixed.c
Comment (pabigot) [reply drivers/regulator/regulator_fixed.c:335]: They eliminated "function defined but not referenced" in cases where only `_sync` regulators are present.

I may be able to remove them with preprocessor conditional that prevent them from being defined, but I think I tried that and it got ugly.

File: drivers/regulator/regulator_fixed.c
Comment (sjg20) [reply drivers/regulator/regulator_fixed.c:284]: Yes I think we still want comments

File: drivers/regulator/regulator_fixed.c
Comment (sjg20) [reply drivers/regulator/regulator_fixed.c:335]: How about __maybe_unused ?

File: tests/drivers/regulator/fixed/README.txt
Comment (sjg20) [reply tests/drivers/regulator/fixed/README.txt:26]: OK

File: boards/arm/thingy52_nrf52832/thingy52_nrf52832.dts
Comment (sjg20) [reply boards/arm/thingy52_nrf52832/thingy52_nrf52832.dts:38]: ok

File: drivers/regulator/regulator_fixed.c
Comment (pabigot) [reply drivers/regulator/regulator_fixed.c:335]: Possibly; no precedent in zephyr.

**Review by dcpleung — APPROVED** [review]

**Review by wentongwu — APPROVED** [review]

File: dts/bindings/regulator/regulator-fixed.yaml
Comment (galak) [inline dts/bindings/regulator/regulator-fixed.yaml:24]: Should `regulator-name` move into regulator.yaml as optional.

File: dts/bindings/regulator/regulator-fixed.yaml
Comment (galak) [inline dts/bindings/regulator/regulator-fixed.yaml:18]: Are we trying to be in sync with the linux binding for `regulator-fixed`?

File: dts/bindings/regulator/regulator-fixed.yaml
Comment (pabigot) [reply dts/bindings/regulator/regulator-fixed.yaml:24]: It could, but for now this is the only binding.  If we add non-fixed regulators it'd make sense to refactor for a base.

File: dts/bindings/regulator/regulator-fixed.yaml
Comment (pabigot) [reply dts/bindings/regulator/regulator-fixed.yaml:18]: To a reasonable degree, yes.  We don't support external properties for the active level and open drain, for reasons described in the binding.

File: dts/bindings/regulator/regulator-fixed.yaml
Comment (galak) [inline dts/bindings/regulator/regulator-fixed.yaml:29]: Looks like upstream just calls this `gpio` for this property.

Comment (pabigot) [reply dts/bindings/regulator/regulator-fixed.yaml:29]: Yes, that's true.  upstream uses `gpio`, `enable-active-high`, and `gpio-open-drain`.  This uses `enable-gpio` to handle all three of those.

I though we'd discussed this and you agreed we shouldn't attempt to support the extra properties when proper use of a GPIO selector makes them unnecessary.  Given that, I think we should use a different name, because `gpio` would suggest the flags are ignored.

Comment (pabigot) [reply dts/bindings/regulator/regulator-fixed.yaml:29]: If we need to name this `zephyr,regulator-fixed` to avoid confusion I'm fine with that.

Comment (galak) [reply dts/bindings/regulator/regulator-fixed.yaml:29]: I think we can leave things as is, but can we add a comment in the binding to explain how/why `enable-gpios` differs from `gpio`?

Comment (pabigot) [reply dts/bindings/regulator/regulator-fixed.yaml:29]: @galak Rebased (since this hasn't been tested directly on master since August), and then [updated the text](https://github.com/zephyrproject-rtos/zephyr/compare/577031aaef1a3f6c0f0c54d5611e6093b5c53884..1056280ca4c5348d839fd6ce39816a8c0877376e).

Comment (pabigot) [reply dts/bindings/regulator/regulator-fixed.yaml:29]: I went with documentation text rather than comment, because if we're going to generate human-readable documentation for these things the comments are less visible than when we expect people to look at the binding file directly.

**Review by carlescufi — APPROVED** [review]

**Pull request merged**
